sdk/core/System.BinaryData/src/BinaryData.cs

```diff
@@ -0,0 +1,264 @@
+
+        /// <summary>
+        /// The backing store for the <see cref="BinaryData"/> instance.
+        /// </summary>
+        private readonly ReadOnlyMemory<byte> _bytes;
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by wrapping the
+        /// provided byte array.
+        /// </summary>
+        /// <param name="data">The array to wrap.</param>
+        public BinaryData(byte[] data)
+        {
+            _bytes = data;
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by serializing the provided object to JSON
+        /// using <see cref="JsonSerializer"/>.
+        /// </summary>
+        ///
+        /// <param name="jsonSerializable">The object that will be serialized to JSON using
+        /// <see cref="JsonSerializer"/>.</param>
+        /// <param name="options">The options to use when serializing to JSON.</param>
+        /// <param name="type">The type to use when serializing the data. If not specified, <see cref="object.GetType"/> will
+        /// be used to determine the type.</param>
+#pragma warning disable AZC0014 // Avoid using banned types in public API
+        public BinaryData(object jsonSerializable, JsonSerializerOptions? options = default, Type? type = default)
+#pragma warning restore AZC0014 // Avoid using banned types in public API
+        {
+            _bytes = JsonSerializer.SerializeToUtf8Bytes(jsonSerializable, type, options);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by wrapping the
+        /// provided bytes.
+        /// </summary>
+        /// <param name="data">Byte data to wrap.</param>
+        public BinaryData(ReadOnlyMemory<byte> data)
+        {
+            _bytes = data;
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from a string by converting
+        /// the string to bytes using the UTF-8 encoding.
+        /// </summary>
+        /// <param name="data">The string data.</param>
+        public BinaryData(string data)
+        {
+            _bytes = Encoding.UTF8.GetBytes(data);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by wrapping the provided
+        /// <see cref="ReadOnlyMemory{Byte}"/>.
+        /// </summary>
+        /// <param name="data">Byte data to wrap.</param>
+        /// <returns>A wrapper over <paramref name="data"/>.</returns>
+        public static BinaryData FromBytes(ReadOnlyMemory<byte> data) =>
+            new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by wrapping the provided
+        /// byte array.
+        /// </summary>
+        /// <param name="data">The array to wrap.</param>
+        /// <returns>A wrapper over <paramref name="data"/>.</returns>
+        public static BinaryData FromBytes(byte[] data) =>
+            new BinaryData(data);
```

Comment:
Consider if `(byte[])null` should throw ArgumentNullException or work the same as the empty array. (The conversion to ReadOnlyMemory succeeds for null).

Reply:
I'll defer to @KrzysztofCwalina on this one.

Reply:
Personally, I like not throwing here as it gives a simple way to create an empty instance of BinaryData.

Reply:
I don't feel strongly, buy I would rather add .Empty static property if we wanted to provide a simple API to get zero length data. The instance would be a singleton. If people use the ctor, they still create instances of this class.